### PR TITLE
Arjun add budget script

### PR DESCRIPTION
**Pull request by arjun-raja**

### CODE REVIEW

File: Managing/budget_manager.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+// Metrics are written onto output sheet in order stated below. Read the 'Metric'
+// column of the Google Ads documentation to find other metrics to include:
+// https://developers.google.com/adwords/api/docs/appendix/reports/campaign-performance-report
+
+
+// Indices
+
+var CONFIG_HEADER_ROW = 1;
+var CONFIG_DATA_ROW = 3;
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
var CONFIG_FIRST_DATA_ROW = 3;
```

File: Managing/budget_manager.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    Logger.log("clearing sheet");
+    clearSheet(dashboardSheet);
+    Logger.log("Re-downloading budgets");
+    download(spreadsheet);
+    Logger.log("Success");
+}
+
+function download(spreadsheet) {
+
+    var configSheet = spreadsheet.getSheetByName("Configuation");
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
    var configSheet = spreadsheet.getSheetByName("Configuration");
```

File: Managing/budget_manager.js [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        update(spreadsheet);
+    } else {
+        throw ("input for RUN_TYPE variable \"" + runType +
+            "\" not recognised. Please enter either 'DOWNLOAD' or 'UPDATE' (including quotation marks)");
+    }
+}
+
+function update(spreadsheet) {
+
+    var dashboardSheet = spreadsheet.getSheetByName("Dashboard");
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
    var dashboardSheet = spreadsheet.getSheetByName("Budget Dashboard");
```

File: Managing/budget_manager.js [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    clearSheet(dashboardSheet);
+    Logger.log("Re-downloading budgets");
+    download(spreadsheet);
+    Logger.log("Success");
+}
+
+function download(spreadsheet) {
+
+    var configSheet = spreadsheet.getSheetByName("Configuation");
+    var dashboardSheet = spreadsheet.getSheetByName("Dashboard");
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
    var dashboardSheet = spreadsheet.getSheetByName("Budget Dashboard");
```

File: Managing/budget_manager.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    var accountIDIndex = configHeaders.indexOf("Account ID");
+    var accountNameIndex = configHeaders.indexOf("Account Name");
+    var campaignNameContainsIndex = configHeaders.indexOf("Campaign Name Contains");
+    var campaignNameDoesNotContainIndex = configHeaders.indexOf("Campaign Name Doesn't Contain");
+    var startDateIndex = configHeaders.indexOf("Start Date");
+    var endDateIndex = configHeaders.indexOf("End Date");
+
+    //Get all rows of data.
+
+    var allData = configSheet.getRange(CONFIG_DATA_ROW, 1, configSheet.getLastRow() - (CONFIG_HEADER_ROW + 1), configSheet.getLastColumn()).getValues();
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
    var allData = configSheet.getRange(CONFIG_DATA_FIRST_ROW, 1, configSheet.getLastRow() - (CONFIG_HEADER_ROW + 1), configSheet.getLastColumn()).getValues();
```

File: Managing/budget_manager.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            budget.setAmount(budgetToChange["Set New Budget"]);
+
+        } catch (e) {
+            throw ("Unable to update budgets: " + e)
+        }
+    }
+}
+
+function getSpreadsheet(spreadsheetUrl) {
+    Logger.log('Checking spreadsheet: ' + SPREADSHEET_URL + ' is valid.');
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
    Logger.log('Checking spreadsheet: ' + spreadsheetUrl + ' is valid.');
```

File: Managing/budget_manager.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    if (spreadsheetUrl.replace(/[AEIOU]/g, "X") == "https://docs.google.com/YXXR-SPRXXDSHXXT-XRL-HXRX") {
+        throw ("Problem with " + SPREADSHEET_URL +
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js:166]:
```suggestion
        throw ("Problem with " + spreadsheetUrl +
```

File: Managing/budget_manager.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+            " URL: make sure you've replaced the default with a valid spreadsheet URL."
+        );
+    }
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js:169]: Very nice!!

File: Managing/budget_manager.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    var campaignNameDoesNotContain = campaignNameDoesNotContain.split(',').map(function (item) {
+        return item.trim();
+    });
+    var campaignFilterQueries = makeCampaignFilterStatements(campaignNameContains, campaignNameDoesNotContain, ignorePausedCampaigns);
+    var combinedQueries = combineQueries(dates, campaignFilterQueries);
+    return combinedQueries;
+}
+
+function makeCampaignFilterStatements(campaignNameContains, campaignNameDoesNotContain, ignorePausedCampaigns) {
+    var whereStatement = "WHERE BudgetStatus != REMOVED ";
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]: Does it need to be `"WHERE BudgetStatus != 'REMOVED'"`, with REMOVED as a string?

File: Managing/budget_manager.js [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+    var whereStatementsArray = [];
+
+
+    if (ignorePausedCampaigns) {
+        whereStatement += "AND AssociatedCampaignStatus = ENABLED ";
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]: Here too, does it need to be `'ENABLED'`?

File: Managing/budget_manager.js [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    } else {
+        whereStatement += "AND AssociatedCampaignStatus IN ['ENABLED','PAUSED'] ";
+    }
+
+    for (var i = 0; i < campaignNameDoesNotContain.length; i++) {
+        if (campaignNameDoesNotContain == "") {
+            break;;
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
            break;
```

File: Managing/budget_manager.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+function getBudgetData(queries, accountName) {
+    dataRows = []
+    var fields = ["BudgetName", "BudgetId", "BudgetReferenceCount", "Amount"]
+    for (var i = 0; i < queries.length; i++) {
+        var report = AdsApp.report(
+            "SELECT " + fields.map(function (field) {
+                return field;
+            }).join(',') + " FROM BUDGET_PERFORMANCE_REPORT " + queries[i]
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js:283]: I don't think we need the `map` here, could use:
```
"SELECT" + fields.join(',') + " FROM BUDGET_PERFORMANCE_REPORT " + queries[i]
```

File: Managing/budget_manager.js [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    var budgetsToChange = getBudgetsToChange(dashboardSheet);
+    for (var i = 0; i < budgetsToChange.length; i++) {
+        budgetToChange = budgetsToChange[i];
+        updateBudgetOnGoogleAds(budgetToChange);
+    }
+    Logger.log("clearing sheet");
+    clearSheet(dashboardSheet);
+    Logger.log("Re-downloading budgets");
+    download(spreadsheet);
+    Logger.log("Success");
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]: Do we want to also log, before these, `Logger.log("updating budgets");`?

File: Managing/budget_manager.js [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+//////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+// Metrics
+
+// Metrics are written onto output sheet in order stated below. Read the 'Metric'
+// column of the Google Ads documentation to find other metrics to include:
+// https://developers.google.com/adwords/api/docs/appendix/reports/campaign-performance-report
+
+
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]: Do we still need this section?

File: Managing/budget_manager.js [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+        var budgetData = getBudgetData(combinedQueries, row[accountNameIndex]);
+        var accountCurrencyCode = getAccountCurrencyCode();
+        var accountDataRow = [row[accountNameIndex], row[accountIDIndex]]
+        var dashboardRows = budgetData.map(function (budgetDataRow) {
+            return accountDataRow.concat(budgetDataRow.map(function (field) {
+                return field.value;
+            })).concat([accountCurrencyCode])
+        });
+        Logger.log(dashboardRows)
+        writeRowsOntoSheet(dasboardSheet, dashboardRows);
```
Comment (gideon-brainlabs) [inline Managing/budget_manager.js]:
```suggestion
        writeRowsOntoSheet(dashboardSheet, dashboardRows);
```

**Review by gideon-brainlabs — CHANGES_REQUESTED** [review]
Looks great! I've suggested some changes.

File: Managing/budget_manager.js (hunk 14 above)
Comment (arjun-raja) [reply Managing/budget_manager.js]: no - will remove

File: Managing/budget_manager.js (hunk 13 above)
Comment (arjun-raja) [reply Managing/budget_manager.js]: 👍

File: Managing/budget_manager.js (hunk 9 above)
Comment (arjun-raja) [reply Managing/budget_manager.js]: yup it does!

File: Managing/budget_manager.js (hunk 11 above)
Comment (arjun-raja) [reply Managing/budget_manager.js]: ;__;

File: Managing/budget_manager.js (hunk 12 above)
Comment (arjun-raja) [reply Managing/budget_manager.js:283]: `fields.join(',')` won't work, as we want the fields to still be in an array. `Array.from()` would be neater, but haven't seen the option to use modern Javascript in Google Ads Scripts yet.

**Review by arjun-raja — COMMENTED** [review]
thanks for looking through!

**Review by gideon-brainlabs — APPROVED** [review]

**Pull request merged**